let the user pass arguments as parameters to gitops comments

chmouel opened the pull request:
when doing a /test, /retest or on gitops-comments we can now redefine
the paramers, you can define standard params and custom params.

Format supported for args are:

test=foo bar="foo bar" hello="\"world\""

see documentation on this commit for details

Demo: (the issue of passing arguments on /test has been fixed since that demo was made)


https://github.com/openshift-pipelines/pipelines-as-code/assets/98980/134ab2ed-6039-495c-8d09-82bbe6a76d93


TODO: e2e tests

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

chmouel committed:
pass arguments as parameters to gitops comments

when doing a /test, /retest or on gitops-comments we can now redefine
the paramers, you can define standard params and custom params.

Format supported for argumnets are:

test=foo bar="foo bar" hello="\"world\""

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel committed:
Add doc for passing arguments to GitOps commands

Add documentation for passing arguments to GitOps commands. We moved all the
documentation in a separate file to make it easier to read and find
since it was becoming too big in the `running.md` file.

Add some unittest as well so to make sure the documentation match the
implementation :upside_down_face:

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
/test linters

savitaashture reviewed (commented):
I saw demo and looks good to me :) 
I see PR don't have e2e are you planning to send in a different PR ??

chmouel commented:
Oh you right I forgot... Let me add them first

chmouel committed:
Add e2e for pass-arguments-to-gitops-comments

The feature gets test on on-comment and when using /test explicitly.

We test all different argument parsing, we had to refactor the tests to
accept golden files to use that for comparison which is a bit better
than regexp in some case.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
I have added the e2e tests in another commit, feel free to have a look now..

piyush-garg commented:
/lgtm
